### ISSUE TITLE
How to apply a protein change to a given amino acid sequence to produce a mutant amino acid sequence

### ISSUE THREAD

**Issue by nh13** [biocommons/hgvs#641]
I was thinking it would be in one of the mapping or projection classes, but I couldn't find one.  I don't want to have to rely on transcript sequence database lookups or the like, but simply provide the reference AA sequence and the change to apply.

**Comment by reece** [comment]
I'm reading this as: you'd like to create a p. expression for an arbitrary protein change. Right?

As you pointed out in #640, and discussed in #333, there's poor congruence between the nt and aa data models, so the method will look different for nt and aa variation (unfortunately). 

```
>>> var_p = parse("NP_012345.6:p.Ala22Trp")
>>> var_p.posedit.pos.start
AAPosition(base=22, aa=A, uncertain=False)
>>> var_p.posedit.edit
AASub(ref='', alt='W', uncertain=False, init_met=False)
>>> var_p.posedit.edit.alt = "Q"
>>> str(var_p)
'NP_012345.6:p.Ala22Gln'
```

Is that what you wanted?

#333 would aim to create greater parity between the models, and therefore enable something more like `alt = ""` or `alt = "C"` or `alt = "QQQQQQ"`, with the appropriate del or sub or repeat.

**Comment by nh13** [comment]
Agreed with @kchu02 , I want a method where I can supply the wildtype sequence and it applies the protein or cdna change.  It should do some validation if ref/alt are available, but it can't in many situations.

**Comment by mbiokyle29** [comment]
I have a similar use case but expanded to include the use case of having multiple variants. For example if I have 2 SNP variants in DNA space (in which ever space `g.`/`t.`/`c.`), if we assume that these variants are all within a transcript and within a coding region, there are 2 cases:
- The variants are in different codons
- The variants are in the same codon

If case 1 is true, then you can just convert `_to_p`, and pass around both AA variants, as this defines the new AA sequence. However, if case 2 is true, you essentially need to collapse the 2 DNA level variants into a single AA variant. Ex:
```
Codon: AAG -> Lys
V one:   C -> AAC/Asn
V two:  T  -> ATG/Met
both :  TC -> ATC/Ile
```

So i'd like to apply multiple variants to a DNA sequence, making the full alt sequence and then convert to AA (and get the per AA variants back out)

**Comment by reece** [comment]
@mbiokyle29 : Your use case is actually the inverse of the topic of this issue, which is reverse translation from protein to nucleic acid sequence.

Your issue is actually about what HGVS alleles (see the "[ ]" bullet on https://varnomen.hgvs.org/recommendations/general/, the links from there).

We took a stab at alleles a long time ago, but it presents numerous complexities, including how to handle/warn about overlapping variants, shuffling (including shuffling variants into each other), phasing certain/uncertain, coordinate shifting due to upstream indels. All of these have solutions, but understanding the complexity of this implementation requires a pretty deep understanding of HGVS... and even then there are unknown dragons lurking, I suspect. Meanwhile, I rarely see HGVS alleles in the wild, so it was very unclear that the complexity would be worth the effort.

**Comment by reece** [comment]
This issue was closed by stalebot. It has been reopened to give more time for community review.  See [biocommons coding guidelines](https://biocommons.org/contributing/coding-guidelines/) for stale issue and pull request policies.  This resurrection is expected to be a one-time event.
